notebooks/config.ipynb

```diff
@@ -0,0 +1,1046 @@
+      "    :param key_path: Path through keys to the desired output block.\n",
+      "    :param values_needed: Report complete, missing, and template values.\n",
+      "    :param total: Require rendering of all Jinja2 variables/expressions.\n",
+      "    :param dry_run: Log output instead of writing to output.\n",
+      "    :param stdin_ok: OK to read from ``stdin``?\n",
+      "    :return: The ``dict`` representation of the realized config.\n",
+      "    :raises: UWConfigRealizeError if ``total`` is ``True`` and any Jinja2 variable/expression was not rendered.\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "help(config.realize)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "501da514-a654-4511-928f-b2ad7db102b2",
+   "metadata": {},
+   "source": [
+    "The `input_config` parameter takes a config from a string path, <a href=\"https://docs.python.org/3/library/pathlib.html#pathlib.Path\">Path</a> object, Python `dict`, or UW `Config` objects like the `YAMLConfig` object from the <a href=\"#Getting-Config-Objects\">Getting Config Objects</a> section. The format of the input config can be specified with `input_format`. Configs are written to `stdout` if `output_file` is set to `None` or to a file specified by `output_file` and `output_format`. Providing the formats of input and output config files is optional if their file extensions are `ini`, `nml`, `sh`, or `yaml`.\n",
+    "<!--cell 20-->"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "77873e14-db3c-417d-be7a-2ba12c9a38f6",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{'greeting': 'Hello', 'recipient': 'World'}"
+      ]
+     },
+     "execution_count": 11,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "config.realize(\n",
+    "    input_config=config1,\n",
+    "    output_file=Path('tmp/config1.yaml')\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fd1823c4-6c3d-4d4a-a614-4d1238588bdd",
+   "metadata": {},
+   "source": [
+    "The `realize()` method returns a dict version of the config regardless of input type, and the file is written in the YAML format as indicated by the file extension.\n",
+    "<!--cell 22-->"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "id": "4f237a73-da83-4632-990f-644632b15cd9",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "greeting: Hello\n",
+      "recipient: World\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%bash\n",
+    "cat tmp/config1.yaml"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "902c4863-e772-49c6-88f2-31e1ab76b418",
+   "metadata": {},
+   "source": [
+    "Input and output formats are not required to match. This can be used to convert a config from one format to another.\n",
```

Comment:
```suggestion
    "Input and output formats are not required to match. This can be used to convert some configs from one format to another.\n",
```
This ability depends on depth- and format-compatibility of the configs. For example:
```
(DEV-uwtools) ~/git/uwtools $ cat a.nml
&a n=88 /
(DEV-uwtools) ~/git/uwtools $ uw config realize --input-file a.nml --output-format nml
&a
    n = 88
/
(DEV-uwtools) ~/git/uwtools $ uw config realize --input-file a.nml --output-format yaml
a:
  n: 88
(DEV-uwtools) ~/git/uwtools $ uw config realize --input-file a.nml --output-format ini
[2024-09-23T16:37:46]    ERROR Accepted output formats for input format nml are nml or yaml
```
IIRC we actually only support output to 1. the same format as the input, or 2. YAML, e.g.
```
(DEV-uwtools) ~/git/uwtools $ cat a.ini 
[a]
n = 88
(DEV-uwtools) ~/git/uwtools $ uw config realize --input-file a.ini --output-format ini
[a]
n = 88
(DEV-uwtools) ~/git/uwtools $ uw config realize --input-file a.ini --output-format yaml
a:
  n: '88'
(DEV-uwtools) ~/git/uwtools $ uw config realize --input-file a.ini --output-format nml
[2024-09-23T16:40:17]    ERROR Accepted output formats for input format ini are ini or yaml
```
Oh, and maybe YAML -> anything is ok, provided the YAML depth conforms to the output format.

You might need to check the code to verify this if you think we should be explicit, or we could just hedge here given that the `uwtools` will complain (informatively) if an unsupported conversion is attempted.

In addition to the depth problem, there are issues with expressiveness between the formats. For example, INI and SH configs are untyped: Their values are always `str` in Python:
```
>>> from uwtools.api.config import get_ini_config
>>> c = get_ini_config("a.ini")
>>> type(c["a"]["n"])
<class 'str'>
```
And so it would be dangerous to use these values to populate, say, an NML config where `int` values are expected.

Reply:
I added your suggestion and did my best to capture the limitations in a few sentences in this markdown cell.